### PR TITLE
Feature/holo 612 royalty smart contract improvements

### PR DESCRIPTION
**Pull request by alexanderattar**
### Royalty contract improvements based on feedback from C4 audit:

- [x] https://github.com/code-423n4/2022-10-holograph-findings/issues/33
- [x] https://github.com/code-423n4/2022-10-holograph-findings/issues/164
- [x] https://github.com/code-423n4/2022-10-holograph-findings/issues/180
- [x] https://github.com/code-423n4/2022-10-holograph-findings/issues/186
- [x] https://github.com/code-423n4/2022-10-holograph-findings/issues/377
- [x] https://github.com/code-423n4/2022-10-holograph-findings/issues/384
- [x] https://github.com/code-423n4/2022-10-holograph-findings/issues/451
- [x] https://github.com/code-423n4/2022-10-holograph-findings/issues/454
- [x] https://github.com/code-423n4/2022-10-holograph-findings/issues/455
- [x] https://github.com/code-423n4/2022-10-holograph-findings/issues/456
- [x] https://github.com/code-423n4/2022-10-holograph-findings/issues/476
- [x] https://github.com/code-423n4/2022-10-holograph-findings/issues/490

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] Code styles have been enforced
- [x] All Hardhat tests are passing

### CODE REVIEW

File: src/enforcer/HolographRoyalties.sol [hunk 1]
```diff
@@ -297,18 +297,13 @@ contract HolographRoyalties is Admin, Owner, Initializable {
     address payable[] memory addresses = _getPayoutAddresses();
     uint256[] memory bps = _getPayoutBps();
     uint256 length = addresses.length;
-    // accommodating the 2300 gas stipend
-    // adding 1x for each item in array to accomodate rounding errors
-    uint256 gasCost = (23300 * length) + length;
     uint256 balance = address(this).balance;
-    require(balance - gasCost > 10000, "ROYALTIES: Not enough ETH");
-    balance = balance - gasCost;
+    require(balance > 10000, "ROYALTIES: Not enough ETH");
```
Comment (alexanderattar) [inline src/enforcer/HolographRoyalties.sol]: Should we still keep this check? or at least add a comment why we're limiting payout to be at least 10_000 wei?

File: src/enforcer/HolographRoyalties.sol [hunk 2]
```diff
@@ -324,11 +319,16 @@ contract HolographRoyalties is Admin, Owner, Initializable {
     uint256 balance = erc20.balanceOf(address(this));
     require(balance > 10000, "ROYALTIES: Not enough tokens");
```
Comment (alexanderattar) [inline src/enforcer/HolographRoyalties.sol]: Same thing here

File: src/enforcer/HolographRoyalties.sol [hunk 3]
```diff
@@ -297,18 +297,13 @@ contract HolographRoyalties is Admin, Owner, Initializable {
...
-    // accommodating the 2300 gas stipend
-    // adding 1x for each item in array to accomodate rounding errors
-    uint256 gasCost = (23300 * length) + length;
     uint256 balance = address(this).balance;
-    require(balance - gasCost > 10000, "ROYALTIES: Not enough ETH");
-    balance = balance - gasCost;
+    require(balance > 10000, "ROYALTIES: Not enough ETH");
     uint256 sending;
-    // uint256 sent;
     for (uint256 i = 0; i < length; i++) {
```
Comment (alexanderattar) [inline src/enforcer/HolographRoyalties.sol:323]: Multiple wardens suggested we move payments to a pull instead of push approach. That would require more changes and likely keeping a mapping of payees. Before moving forward with that I'd like to get your input Vitto https://github.com/code-423n4/2022-10-holograph-findings/issues/104

File: src/enforcer/HolographRoyalties.sol (hunk 1 above)
Comment (ACC01ADE) [reply src/enforcer/HolographRoyalties.sol]: We should keep this since if we allow for lower than 10000 WEI, we have a chance that some wallets in payouts won't even get anything due to rounding errors. Plus don't see a point triggering royalties for dust amounts.

File: src/enforcer/HolographRoyalties.sol (hunk 2 above)
Comment (ACC01ADE) [reply src/enforcer/HolographRoyalties.sol]: Need to remove it from here since some tokens are using no decimals or two decimal places, this would result in $100 for something like USDT getting stuck in account. I think there was a specific C4 issues associated with this one.

File: src/enforcer/HolographRoyalties.sol (hunk 1 above)
Comment (alexanderattar) [reply src/enforcer/HolographRoyalties.sol]: sounds good

Comment (alexanderattar) [reply src/enforcer/HolographRoyalties.sol:301]: I removed

File: src/enforcer/HolographRoyalties.sol (hunk 2 above)
Comment (alexanderattar) [reply src/enforcer/HolographRoyalties.sol]: I removed this one

File: src/enforcer/HolographRoyalties.sol (hunk 3 above)
Comment (ACC01ADE) [reply src/enforcer/HolographRoyalties.sol:323]: Our goal with the HolographRoyalties contract is to only collect the royalties and provide a way to access the amounts after. If only one payout recipient is configured, this could be considered as a pull approach. The push approach is set as the default/only option simply due to the fact that we do not plan to add any features/functionality outside of the scope of collecting the royalties. Any additional distribution logic should be handled by an external contract code.

File: contracts/enforcer/HolographRoyalties.sol [hunk 4]
```diff
@@ -696,4 +749,31 @@ contract HolographRoyalties is Admin, Owner, Initializable {
   function getTokenAddress(string memory tokenName) public view returns (address) {
     return _getTokenAddress(tokenName);
   }
+
+  /**
+   * @notice Used to wrap function calls to check if they return without revert regardless of return type.
+   * @dev
```
Comment (alexanderattar) [inline contracts/enforcer/HolographRoyalties.sol]: @ACC01ADE Would you be able to add an extra note here on the logic. I know you mentioned it while we were chatting yesterday, but wasn't sure the best way to articulate it in writing

File: src/enforcer/HolographRoyalties.sol [hunk 5]
```diff
@@ -597,4 +650,31 @@ contract HolographRoyalties is Admin, Owner, Initializable {
   function getTokenAddress(string memory tokenName) public view returns (address) {
     return _getTokenAddress(tokenName);
   }
+
+  /**
+   * @notice Used to wrap function calls to check if they return without revert regardless of return type.
+   * @dev
+   * @return Returns true if the wrapped function call returns without a revert even if it doesn't return true.
+   */
+  function _callOptionalReturn(
```
Comment (ACC01ADE) [inline src/enforcer/HolographRoyalties.sol:668]: In order to accommodate non-compliant ERC-20 tokens (like USDT), which do not return a bool on success, we wrap the ERC20 calls. This allows to not revert when a call returns no data. The logic works by: making a call, checking if return opcode is a revert, if it is then it reverts as well, if it's not then it checks for return data, if return data exists, it is returned as a bool, if return data does not exist (0 length) then success is expected and returns true.

File: contracts/enforcer/HolographRoyalties.sol (hunk 4 above)
Comment (ACC01ADE) [reply contracts/enforcer/HolographRoyalties.sol]: Replied in the `src` version of the smart contract. [Here](https://github.com/holographxyz/holograph-protocol/pull/93#discussion_r1038542139)

**Comment by ACC01ADE** [comment]
@alexanderattar I think this is all perfect. One note is that we should add the restriction of maximum array length of 10 for `configurePayouts` in order to avoid users from making configurations that would result in permanent out of gas errors in payout for loops.

**Comment by sogoiii** [comment]
@alexanderattar How are you resolving this issue? https://github.com/code-423n4/2022-10-holograph-findings/issues/451 I see that you removed a bunch of checks.

File: test/20_royalties_tests.ts [hunk 6]
```diff
@@ -540,4 +550,216 @@ describe('HolographRoyalties Contract', async function () {
...
+      const totalRoyalties = ethers.utils.parseEther('1');
+
+      before(async () => {
+        const data = (await royalties.populateTransaction.configurePayouts([owner.address], [10000])).data || '';
+        const tx = await l1.factory.connect(owner).adminCall(royalties.address, data);
+        await tx.wait();
+      });
+
+      it('should be able to withdraw all native token balance', async () => {
+        await owner.sendTransaction({
```
Comment (sogoiii) [inline test/20_royalties_tests.ts:567]: @alexanderattar I think this should not come from the owner but from somewhere else.

File: test/20_royalties_tests.ts [hunk 7]
```diff
@@ -540,4 +550,216 @@ describe('HolographRoyalties Contract', async function () {
...
+          to: royalties.address,
+          value: totalRoyalties,
+        });
+
+        const accountBalanceBefore = await ethers.provider.getBalance(owner.address);
+        const contractBalanceBefore = await ethers.provider.getBalance(royalties.address);
```
Comment (sogoiii) [inline test/20_royalties_tests.ts:573]: We should have an expect check for `contractBalanceBefore` that is tied to the `totalRoyalties` balance

File: test/20_royalties_tests.ts [hunk 8]
```diff
@@ -540,4 +550,216 @@ describe('HolographRoyalties Contract', async function () {
...
+
+        const data = (await royalties.populateTransaction.getEthPayout()).data || '';
+        const tx = await l1.factory.connect(owner).adminCall(royalties.address, data);
+        await tx.wait();
+
+        const accountBalanceAfter = await ethers.provider.getBalance(owner.address);
+        const contractBalanceAfter = await ethers.provider.getBalance(royalties.address);
+
+        expect(contractBalanceAfter).to.be.lt(contractBalanceBefore);
```
Comment (sogoiii) [inline test/20_royalties_tests.ts:582]: We can keep these 2 tests. But I think we should be able to create an expect with a specific value.

File: test/20_royalties_tests.ts [hunk 9]
```diff
@@ -540,4 +550,216 @@ describe('HolographRoyalties Contract', async function () {
...
+    describe('A collection with 1 recipient', async () => {
+      const totalRoyalties = ethers.utils.parseEther('1');
+
+      before(async () => {
+        const data = (await royalties.populateTransaction.configurePayouts([owner.address], [10000])).data || '';
+        const tx = await l1.factory.connect(owner).adminCall(royalties.address, data);
+        await tx.wait();
+      });
+
+      it('should be able to withdraw all native token balance', async () => {
```
Comment (sogoiii) [inline test/20_royalties_tests.ts:566]: I think we need to add a test for the zero case. You guys mentioned something about this in the chat.

File: test/20_royalties_tests.ts [hunk 10]
```diff
@@ -540,4 +550,216 @@ describe('HolographRoyalties Contract', async function () {
...
+
+        const data = (await royalties.populateTransaction.getTokenPayout(ERC20.address)).data || '';
+        const tx = await l1.factory.connect(owner).adminCall(royalties.address, data);
+        await tx.wait();
+
+        const accountBalanceAfter = await ERC20.balanceOf(owner.address);
+        const contractBalanceAfter = await ERC20.balanceOf(royalties.address);
+
+        expect(contractBalanceAfter).to.be.lt(contractBalanceBefore);
+        expect(accountBalanceAfter).to.be.gt(accountBalanceBefore);
```
Comment (sogoiii) [inline test/20_royalties_tests.ts:604]: Again, we should be able to check specific values.

File: test/20_royalties_tests.ts [hunk 11]
```diff
@@ -540,4 +550,216 @@ describe('HolographRoyalties Contract', async function () {
...
+
+        const accountABalanceAfter = await ethers.provider.getBalance(anyAddress);
+        const accountBBalanceAfter = await ethers.provider.getBalance(notOwner.address);
+        const contractBalanceAfter = await ethers.provider.getBalance(royalties.address);
+
+        const sixtyPercentOfRoyalties = contractBalanceBefore.sub(contractBalanceAfter).mul(60).div(100);
+        const fortyPercentOfRoyalties = contractBalanceBefore.sub(contractBalanceAfter).mul(40).div(100);
+
+        expect(contractBalanceAfter).to.be.lt(contractBalanceBefore);
+        expect(Number(accountABalanceAfter)).to.be.approximately(
```
Comment (sogoiii) [inline test/20_royalties_tests.ts:641]: We should add a note to describe why we use approximately here. I think we are using because of the minor differences in computing percentages. But because we define the percentage split above, we should be able to know the exact amount.

File: test/20_royalties_tests.ts [hunk 12]
```diff
@@ -540,4 +550,216 @@ describe('HolographRoyalties Contract', async function () {
...
+        const accountBBalanceAfter = await ethers.provider.getBalance(notOwner.address);
+        const accountCBalanceAfter = await ethers.provider.getBalance(mockAddress);
+        const contractBalanceAfter = await ethers.provider.getBalance(royalties.address);
+
+        const twentyPercentOfRoyalties = contractBalanceBefore.sub(contractBalanceAfter).mul(20).div(100);
+        const fiftyPercentOfRoyalties = contractBalanceBefore.sub(contractBalanceAfter).mul(50).div(100);
+        const thirtyPercentOfRoyalties = contractBalanceBefore.sub(contractBalanceAfter).mul(30).div(100);
+
+        expect(contractBalanceAfter).to.be.lt(contractBalanceBefore);
+        expect(Number(accountABalanceAfter)).to.be.approximately(
```
Comment (sogoiii) [inline test/20_royalties_tests.ts:720]: We should add a note to describe why we use approximately here. I think we are using because of the minor differences in computing percentages. But because we define the percentage split above, we should be able to know the exact amount.

File: test/20_royalties_tests.ts (hunk 9 above)
Comment (alexanderattar) [reply test/20_royalties_tests.ts:566]: Added this

File: test/20_royalties_tests.ts (hunk 6 above)
Comment (alexanderattar) [reply test/20_royalties_tests.ts:567]: where did you have in mind? I think the idea here is it's just funding the contract so there's royalties to pull out. owner is actually `owner = l1.deployer;`. we could change it to be from another account though if it makes more sense

File: test/20_royalties_tests.ts (hunk 11 above)
Comment (alexanderattar) [reply test/20_royalties_tests.ts:641]: Not going to use approximately anymore now that the contract logic has been fixed

File: test/20_royalties_tests.ts (hunk 12 above)
Comment (alexanderattar) [reply test/20_royalties_tests.ts:720]: See above

File: test/20_royalties_tests.ts (hunk 10 above)
Comment (alexanderattar) [reply test/20_royalties_tests.ts:604]: yep updated to use equals instead

File: test/20_royalties_tests.ts (hunk 8 above)
Comment (alexanderattar) [reply test/20_royalties_tests.ts:582]: Using equals now instead

File: test/20_royalties_tests.ts (hunk 7 above)
Comment (alexanderattar) [reply test/20_royalties_tests.ts:573]: Done

**Comment by ACC01ADE** [comment]
Made two minor edits/improvements. At this point I believe the PR is ready. We can make any additional tests and final edits part of a new PR (if needed).

**Review by ACC01ADE — APPROVED** [review]
